File: daily_temps.rs
struct Solution;

impl Solution {
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let n = temperatures.len();
        let mut ans: Vec<i32> = vec![0; n];
        let mut stk: Vec<usize> = Vec::new();
        for i in 0..n {
            while stk.last().map_or(false, |&v| { temperatures[v] < temperatures[i]}) {
                match stk.pop() {
                    None => break,
                    Some(idx) => ans[idx] = (i - idx) as i32
                }
            }
            stk.push(i);
        }
        ans
    }
}

fn main() {
    let t1 = vec![73,74,75,71,69,72,76,73];
    let t2 = vec![30,40,50,60];
    let t3 = vec![30,60,90];
    println!("{:?}", Solution::daily_temperatures(t1));
    println!("{:?}", Solution::daily_temperatures(t2));
    println!("{:?}", Solution::daily_temperatures(t3));
}

File: other/daily_temps.cpp
#include "leetcode.h"

class Solution {
public:
  vector<int> dailyTemperatures(vector<int> &temperatures) {
    int n = temperatures.size();
    vector<int> nge(n, 0);
    stack<int> stk{};
    for (int i = n - 1; i >= 0; --i) {
      while (!stk.empty() && temperatures[stk.top()] <= temperatures[i])
        stk.pop();
      if (!stk.empty())
        nge[i] = stk.top() - i;
      stk.push(i);
    }
    return nge;
  }
};

int main() {
  Solution obj;
  vector<int> t1 = {73, 74, 75, 71, 69, 72, 76, 73};
  vector<int> t2 = {30, 40, 50, 60};
  vector<int> t3 = {30, 60, 90};
  obj.dailyTemperatures(t1); // expect 1,1,4,2,1,1,0,0
  obj.dailyTemperatures(t2); // expect 1,1,1,0
  obj.dailyTemperatures(t3); // expect 1,1,0
}
